llvm/test/CodeGen/NVPTX/param-add.ll

```diff
@@ -0,0 +1,44 @@
+; RUN: llc < %s -march=nvptx64 --debug-counter=dagcombine=0 | FileCheck %s
+; RUN: %if ptxas %{ llc < %s -march=nvptx64 | %ptxas-verify %}
+
+%struct.8float = type <{ [8 x float] }>
+
+declare i32 @callee(%struct.8float %a)
+
+define i32 @test(%struct.8float alignstack(32) %data) {
+  ;CHECK-NOT: add.
+  ;CHECK-DAG: ld.param.u8 %r{{.*}}, [test_param_0];
+  ;CHECK-DAG: ld.param.u8 %r{{.*}}, [test_param_0+1];
```

Comment:
We're already lowering loads as `[symbol+offset]` without the patch: https://godbolt.org/z/zWsnvqPjd

Can you add a test that demonstrating what this patch is intended to change?

Reply:
I don't think `--debug-counter=dagcombine=0` has any effect in release builds. Here's what I get for llc (assertions trunk): https://godbolt.org/z/95fx4nMxq. An illegal `add` is on line 25 of the output.

Reply:
Whoa. We get different output with assertions enabled? Now, *that* does sound like a bug to me.

If an assertion fails somewhere the compilation should've failed. We need to track down the source of this divergence.
Making back-end changes to avoid `add symbol, const` may be just hiding the problem. It's also something to be fixed, but the divergence in the assertions-enabled build is a more severe problem, IMO.

Reply:
In this case I was purposefully creating a divergence by disabling DAG combines through a debug counter, which seems to be impossible in release mode NVPTX. It's a somewhat artificial case designed to demonstrate that the backend is incorrect without DAG combines. I'm not sure whether or not there is some other divergence going on here...does disabling combines explain it or am I misunderstanding something?

Reply:
> Whoa. We get different output with assertions enabled? Now, that does sound like a bug to me.

When assertions aren't enabled (I believe) [`debug-counter` does not have any effect](https://github.com/llvm/llvm-project/blob/38099d0608342ddff0737a048ca5fa325c4b0749/llvm/include/llvm/Support/DebugCounter.h#L154-L156).

> We need to track down the source of this divergence.

I'd also be interested to see which optimization is eliminating the `add symbol, const`.

Reply:
It appears that the differences start at the very beginning, in the `nvptx-isel` https://godbolt.org/z/Kdesca3na

Looks like the assert-enabled-build starts with a very different DAG:
![image](https://github.com/user-attachments/assets/fffc3c87-dc50-433b-8333-4526b060ab7c)

Reply:
Without the `-debug-counter=dagcombine=0` these two functions look the same. I can't figure out how to inspect the DAG in a release build so I can't say more there. I agree that a diff in the initial DAGs would be odd, but I'm not seeing evidence of that from this example.

Reply:
llc `--print-after nvptx-isel` would give you the DAG printout above. Or you can build `llc` with debug info enabled and check what's going on when we're lowering the function arguments.

I find it very useful to run `llc` under `rr` (https://rr-project.org/) which allows you to step back and forth in time so you can backtrack things. E.g. set breakpoint on a node constructor for the `LD_i32_ari_64` and then go back up the stack and possibly back in time in the stack frame where you suspect things may have gone wrong.

Reply:
>llc --print-after nvptx-isel would give you the DAG printout above.

I forgot DAG can refer to post-ISEL machine functions, not just the SelectionDAG, so that makes sense.

> Or you can build llc with debug info enabled and check what's going on when we're lowering the function arguments.

I did some of this with `-debug-only=isel` to inspect the SelectionDAG with and without combines, which is how I figured out this param issue existed.

> I find it very useful to run llc under rr (https://rr-project.org/) which allows you to step back and forth in time so you can backtrack things.

Thanks for sharing this, seems like a good resource!

I'm still confused as to what the problematic divergence is supposed to be, or what I should be trying to figure out. While the DAGs in https://github.com/llvm/llvm-project/pull/119935#discussion_r1887660119 are different, they are printed at a point after DAG combines would have occurred, so it seems reasonable that disabling combines causes a difference. That the same inputs and same flags lead to different outputs also seems reasonable given one of the flags is intended for debugging and is ignored by release builds. If there was a difference in the initial SelectionDAGs then that would be a problem, but I'm not seeing evidence of that.

@justinfargnoli -
> I'd also be interested to see which optimization is eliminating the add symbol, const.

`add symbol, const` is not eliminated in an optimization. The relevant optimization is something like `add (add symbol, const1), const2 -> add symbol, const3` where `const3=const1+const2`. The `add symbol, const` nodes/instructions are supposed to end up unused during ISEL and thus be removed from the DAG. The `SelectADDRsi` functions examine those nodes to find the symbol and offset and use those directly in loads/stores. The issue arises when they are used in other instructions than loads or stores (like adds) as these don't have the selection logic to deal with them properly. This change isn't really porting a DAG combine, it's making the `SelectADDRsi` functions better able to fulfill their purpose of finding a symbol and offset for use in a load/store.

Reply:
> I'm still confused as to what the problematic divergence is supposed to be, 

In general, enabling/disabling assertions is not supposed to alter LLVM's output, assuming that all assertions succeed.
The fact that we have the output changing means that no-assertion build may be relying on some UB that gets affected by the additional checks that are supposed to have no effect other than validating some invariant.

While it's possible to intentionally make code behave with/without assertions, I can't think of a good reason why it would be the case here.

Reply:
> I can't think of a good reason why it would be the case here

Could you explain why the debug counter is not a good reason for the different behavior?